src/modules/mc_att_control/mc_att_control_main.cpp

```diff
@@ -542,75 +541,15 @@ MulticopterAttitudeControl::control_attitude()
 	// physical thrust axis is the negative of body z axis
 	_thrust_sp = -_v_att_sp.thrust_body[2];
 
-	/* prepare yaw weight from the ratio between roll/pitch and yaw gains */
-	Vector3f attitude_gain = _attitude_p;
-	const float roll_pitch_gain = (attitude_gain(0) + attitude_gain(1)) / 2.f;
-	const float yaw_w = math::constrain(attitude_gain(2) / roll_pitch_gain, 0.f, 1.f);
-	attitude_gain(2) = roll_pitch_gain;
-
-	/* get estimated and desired vehicle attitude */
-	Quatf q(_v_att.q);
-	Quatf qd(_v_att_sp.q_d);
-
-	/* ensure input quaternions are exactly normalized because acosf(1.00001) == NaN */
-	q.normalize();
-	qd.normalize();
-
-	/* calculate reduced desired attitude neglecting vehicle's yaw to prioritize roll and pitch */
-	Vector3f e_z = q.dcm_z();
-	Vector3f e_z_d = qd.dcm_z();
-	Quatf qd_red(e_z, e_z_d);
-
-	if (abs(qd_red(1)) > (1.f - 1e-5f) || abs(qd_red(2)) > (1.f - 1e-5f)) {
-		/* In the infinitesimal corner case where the vehicle and thrust have the completely opposite direction,
-		 * full attitude control anyways generates no yaw input and directly takes the combination of
-		 * roll and pitch leading to the correct desired yaw. Ignoring this case would still be totally safe and stable. */
-		qd_red = qd;
+	if ((_v_control_mode.flag_control_velocity_enabled || _v_control_mode.flag_control_auto_enabled) &&
+		!_v_control_mode.flag_control_manual_enabled) {
+		_attitude_control.setRateLimit(_auto_rate_max);
```

Comment:
Don't we only need to update the rate limit on control mode change?

Reply:
In principle yes, I now first only made gains only update when they change but the plan clearly is to improve it further.

Reply:
I addressed this comment in https://github.com/PX4/Firmware/pull/11308/commits/71b9d69c123b4cce8304343a5f58c1b291c7542c

I think this extra logic should soon be removed completely. The contraints get part of the controller interface and are received together with the setpoint for example like here: https://github.com/PX4/Firmware/blob/8947cb5841acae5de58716b8933525851db975a9/src/modules/mc_pos_control/mc_pos_control_main.cpp#L742-L749 the generator of the attitude setpoints e.g. the position controller should decide what limits go well with the setpoints it publishes.